File: 2121-find-if-path-exists-in-graph/2121-find-if-path-exists-in-graph.py
class Solution(object):
    def validPath(self, n, edges, source, destination):
        """
        :type n: int
        :type edges: List[List[int]]
        :type source: int
        :type destination: int
        :rtype: bool
        """
        
        # Create the graph
        graph = {}
        for i in range(n):
            graph[i] = []
        
        for u, v in edges:
            graph[u].append(v)
            graph[v].append(u)
        
        # Set to keep track of visited nodes
        visited = set()
        
        # Define the recursive DFS function
        def dfs(node):
            # If we've reached the destination, we're done
            if node == destination:
                return True
            
            # Mark the current node as visited
            visited.add(node)
            
            # Check all neighbors
            for neighbor in graph[node]:
                if neighbor not in visited:
                    if dfs(neighbor):
                        return True
            
            # If we get here, this path didn't work
            return False
        
        # Start the DFS from the source
        return dfs(source)

File: 2121-find-if-path-exists-in-graph/2121-find-if-path-exists-in-graph.ts
function validPath(n: number, edges: number[][], source: number, destination: number): boolean {
    let graph: { [key: number]: number[] } = {};

    // Initialize the graph
    for (let i = 0; i < n; i++) {
        graph[i] = [];
    }

    // Populate the graph
    for (let edge of edges) {
        let u = edge[0];
        let v = edge[1];

        graph[u].push(v);
        graph[v].push(u);
    }

    const visited: Set<number> = new Set(); // Use a Set for visited nodes

    return dfs(graph, source, destination, visited);
}

function dfs(graph: { [key: number]: number[] }, source: number, destination: number, visited: Set<number>): boolean {
    if (source === destination) {
        return true;
    }

    visited.add(source); // Mark the current node as visited

    for (let neighbor of graph[source]) {
        if (!visited.has(neighbor)) { // Check using Set
            if (dfs(graph, neighbor, destination, visited)) {
                return true;
            }
        }
    }

    return false;
}
